### PR TITLE
feat(wren-ai-service): integrate dspy into wrenAI

### PR DESCRIPTION
**Pull request by tedyyan**
- Generate evaluation dataset
[[wren-ai-service/eval/prompt_optimizer.py](https://github.com/Canner/WrenAI/blob/main/wren-ai-service/eval/README.md#eval-dataset-preparationif-using-spider-10-dataset)](https://github.com/Canner/WrenAI/blob/main/wren-ai-service/eval/README.md#eval-dataset-preparationif-using-spider-10-dataset)

======================= 1 =====================
================ Build the module =================

use eval.py and the spider2 v1 dataset to train and optimize a dspy module (https://github.com/taoyds/spider/tree/master/evaluation_examples/examples)
The prediction_eval_ask_9df57d69-250c-4a10-b6a5-6595509fed6b_2024_10_23_132136.toml is a predict dataset without dspy

```
wren-ai-service/eval/prompt_optimizer.py --training-dataset spider_car_1_eval_dataset.toml --file prediction_eval_ask_9df57d69-250c-4a10-b6a5-6595509fed6b_2024_10_23_132136.toml
```

output: `eval/optimized/AskGenerationV1_optimized_2024_10_21_181426.json`

====================== 2 =======================
=================Prediction =================

1. put the trained dspy module into an environment variable `DSPY_OPTIMAZED_MODEL`

```
export DSPY_OPTIMAZED_MODEL=eval/optimized/AskGenerationV1_optimized_2024_10_21_181426.json
```

2. start predict get the predicted result

```
just predict eval/dataset/spider_car_1_eval_dataset.toml
```

3. output

```
outputs/predictions/prediction_eval_ask_f5103405-09b2-448c-829d-cedd3c3b12d0_2024_10_22_184950.toml

```

====================== 3 =======================
=============== Evaluate =========================

1. Start to evaluate according to the predicted result

```
just eval prediction_eval_ask_f5103405-09b2-448c-829d-cedd3c3b12d0_2024_10_22_184950.toml

```
image

Notes:
wren-ai-service/eval/prompt_optimizer.py can be improved by incorporating additional training examples or use other modules in dspy

Thanks for @cyyeh guide and codes, most of the PR's code are borrowed from his branch `chore/ai-service/dspy`

### CODE REVIEW

File: wren-ai-service/src/pipelines/generation/sql_generation.py [hunk 1]
```diff
@@ -137,13 +143,32 @@ def prompt(
...
+def configure_llm_provider(llm: str, api_key: str) -> None:
+    dspy.settings.configure(lm=dspy.OpenAI(model=llm, api_key=api_key,max_tokens=4096))
 
 
 @async_timer
 @observe(as_type="generation", capture_input=False)
-async def generate_sql(prompt: dict, generator: Any) -> dict:
-    logger.debug(f"prompt: {orjson.dumps(prompt, option=orjson.OPT_INDENT_2).decode()}")
-    return await generator.run(prompt=prompt.get("prompt"))
+async def generate_sql(prompt: dict, generator: Any, dspy_module: Any) -> dict:
```
Comment (cyyeh) [inline wren-ai-service/src/pipelines/generation/sql_generation.py:160]: @tedyyan could you help figure out if dspy_module supports async?

File: wren-ai-service/src/pipelines/generation/sql_generation.py [hunk 2]
```diff
@@ -178,14 +203,22 @@ class GenerationResults(BaseModel):
...
 
-
 class SQLGeneration(BasicPipeline):
     def __init__(
         self,
         llm_provider: LLMProvider,
         engine: Engine,
         **kwargs,
     ):
+        self.dspy_module = None
```
Comment (cyyeh) [inline wren-ai-service/src/pipelines/generation/sql_generation.py:213]: sorry, could you explain to me what do these lines of code mean?

**Review by cyyeh — COMMENTED** [review]
@tedyyan thanks for your contribution. I think we can schedule a time to discuss this pr

**Comment by tedyyan** [comment]
poetry lock —no-update

**Review by cyyeh — APPROVED** [review]
LGTM, thanks for your contribution

**Pull request merged**